chain/client/src/sync/state.rs

```diff
@@ -800,6 +727,245 @@ impl StateSync {
+                        status: ShardSyncStatus::StateDownloadComplete,
+                    }
+                }
+                Err(err) => {
+                    // Cannot finalize the downloaded state.
+                    // The reasonable behavior here is to start from the very beginning.
+                    tracing::error!(target: "sync", %shard_id, %sync_hash, ?err, "State sync finalizing error");
+                    *shard_sync_download = ShardSyncDownload::new(now);
+                    let shard_state_header = chain.get_state_header(shard_id, sync_hash)?;
+                    let state_num_parts =
+                        get_num_state_parts(shard_state_header.state_root_node().memory_usage);
+                    chain.clear_downloaded_parts(shard_id, sync_hash, state_num_parts)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn sync_shards_download_complete_status(
+        &mut self,
+        split_states: bool,
+        shard_id: ShardId,
+        shard_sync_download: &mut ShardSyncDownload,
+        sync_hash: CryptoHash,
+        chain: &mut Chain,
+    ) -> Result<bool, near_chain::Error> {
+        let shard_state_header = chain.get_state_header(shard_id, sync_hash)?;
+        let state_num_parts =
+            get_num_state_parts(shard_state_header.state_root_node().memory_usage);
+        chain.clear_downloaded_parts(shard_id, sync_hash, state_num_parts)?;
+
+        let mut this_done = false;
+        // If the shard layout is changing in this epoch - we have to apply it right now.
+        if split_states {
+            *shard_sync_download = ShardSyncDownload {
+                downloads: vec![],
+                status: ShardSyncStatus::StateSplitScheduling,
+            }
+        } else {
+            // If there is no layout change - we're done.
+            *shard_sync_download =
+                ShardSyncDownload { downloads: vec![], status: ShardSyncStatus::StateSyncDone };
+            this_done = true;
+        }
+        Ok(this_done)
+    }
+
+    fn sync_shards_state_split_scheduling_status(
+        &mut self,
+        shard_id: ShardId,
+        shard_sync_download: &mut ShardSyncDownload,
+        sync_hash: CryptoHash,
+        chain: &mut Chain,
+        state_split_scheduler: &dyn Fn(StateSplitRequest),
+        me: &Option<AccountId>,
+    ) -> Result<(), near_chain::Error> {
+        let status = Arc::new(StateSplitApplyingStatus::new());
+        chain.build_state_for_split_shards_preprocessing(
+            &sync_hash,
+            shard_id,
+            state_split_scheduler,
+            status.clone(),
+        )?;
+        tracing::debug!(target: "sync", %shard_id, %sync_hash, ?me, "State sync split scheduled");
+        *shard_sync_download = ShardSyncDownload {
+            downloads: vec![],
+            status: ShardSyncStatus::StateSplitApplying(status),
+        };
+        Ok(())
+    }
+
+    /// Returns `this_done`.
```

Comment:
nit: `this_done` could be elaborated more in the comment
(What is "this"? What do you mean by done? If the chain is done we can shut down the node, right? :P )